### PR TITLE
[mlir][bufferization] Fix OneShotBufferize when `defaultMemorySpaceFn` is used

### PR DESCRIPTION
**Pull request by christopherbate**
As described in issue llvm/llvm-project#91518, a previous PR
llvm/llvm-project#78484 introduced the `defaultMemorySpaceFn` into
bufferization options, allowing one to inform OneShotBufferize that it
should use a specified function to derive the memory space attribute
from the encoding attribute attached to tensor types.

However, introducing this feature exposed unhandled edge cases,
examples of which are introduced by this change in the new test under
`test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir`.

Fixing the inconsistencies introduced by `defaultMemorySpaceFn` is pretty
simple. This change:

- Updates the `bufferization.to_memref` and `bufferization.to_tensor`
  operations to explicitly include operand and destination types,
  whereas previously they relied on type inference to deduce the
  tensor types. Since the type inference cannot recover the correct
  tensor encoding/memory space, the operand and result types must be
  explicitly included. This is a small assembly format change, but it
  touches a large number of test files.

- Makes minor updates to other bufferization functions to handle the
  changes in building the above ops.

- Updates bufferization of `tensor.from_elements` to handle memory
  space.

Integration/upgrade guide:

In downstream projects, if you have tests or MLIR files that explicitly use
`bufferization.to_tensor` or `bufferization.to_memref`, then update
them to the new assembly format as follows:

```
%1 = bufferization.to_memref %0 : memref<10xf32>
%2 = bufferization.to_tensor %1 : memref<10xf32>
```

becomes

```
%1 = bufferization.to_memref %0 : tensor<10xf32> to memref<10xf32>
%2 = bufferization.to_tensor %0 : memref<10xf32> to tensor<10xf32> 
```

### CODE REVIEW

**Comment by christopherbate** [comment]
@matthias-springer Can you review? The vast majority of changes here are just to update tests due to change in assembly formats for `bufferization.to_tensor|to_memref`.

File: mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td [hunk 1]
```diff
@@ -496,7 +501,8 @@ def Bufferization_ToMemrefOp : Bufferization_Op<"to_memref", [
     Pure,
     TypesMatchWith<"type of 'tensor' is the tensor equivalent of 'memref'",
                    "memref", "tensor",
-                   "memref::getTensorTypeFromMemRefType($_self)">
+                   "memref::getTensorTypeFromMemRefType($_self)",
+                   "bufferization::detail::tensorTypesMatchUpToEncoding">
```
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td:505]: I think you can use `AllShapesMatch` and `AllElementTypesMatch` instead of `tensorTypesMatchUpToEncoding`.

File: mlir/lib/Dialect/Bufferization/Transforms/Bufferize.cpp [hunk 2]
```diff
@@ -67,10 +67,14 @@ BufferizeTypeConverter::BufferizeTypeConverter() {
                               ValueRange inputs, Location loc) -> Value {
     assert(inputs.size() == 1 && "expected exactly one input");
 
+    // Unranked to ranked casts must be explicit.
+    if (auto inputType = dyn_cast<UnrankedMemRefType>(inputs[0].getType()))
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/Bufferize.cpp:71]: Why is this needed?

File: mlir/test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir [hunk 3]
```diff
@@ -0,0 +1,133 @@
+// RUN: mlir-opt %s -one-shot-bufferize="use-encoding-for-memory-space" -split-input-file | FileCheck %s
+
+// TODO: move to tensor dialect tests
```
Comment (matthias-springer) [inline mlir/test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir:3]: Can you address this TODO?

File: mlir/test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+//       CHECK:     %[[v0:.+]] = bufferization.to_memref %[[arg0]] : tensor<128xf32> -> memref<128xf32, strided<[?], offset: ?>>
+//       CHECK:     %[[alloc:.+]] = memref.alloc() {alignment = 64 : i64} : memref<128xf32, 1>
+//       CHECK:     memref.copy %[[v0]], %[[alloc]] : memref<128xf32, strided<[?], offset: ?>> to memref<128xf32, 1>
+//       CHECK:     %[[v1:.+]] = bufferization.to_tensor %[[alloc]] : memref<128xf32, 1> -> tensor<128xf32>
+//       CHECK:     return %[[v1]] : tensor<128xf32>
+
+// -----
+
+func.func @alloc_tesor_copy_from_non_default_space(%arg0: tensor<128xf32, 1>) -> tensor<128xf32, 2> {
+  %0 = bufferization.alloc_tensor() copy(%arg0) {memory_space = 2 : i64} : tensor<128xf32, 1>
```
Comment (matthias-springer) [inline mlir/test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir:81]: What is the meaning of this op? The two memory spaces here are inconsistent.

File: mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td [hunk 5]
```diff
@@ -388,9 +388,7 @@ def Bufferization_ToTensorOp : Bufferization_Op<"to_tensor", [
     BufferizableOpInterface,
     SameOperandsAndResultShape,
     SameOperandsAndResultElementType,
-    TypesMatchWith<"result type matches tensor equivalent of 'memref'",
-                   "memref", "result",
-                   "memref::getTensorTypeFromMemRefType($_self)">
+    AllElementTypesMatch<["memref", "result"]>
   ]> {
   let summary = "create a tensor from a `memref`";
   let description = [{
```
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td:394]: Code examples should be updated in `description`.

**Review by matthias-springer — COMMENTED** [review]
Can you add extra unit tests for `scf.for` (with tensor iter_arg that has an encoding), `scf.if` and `scf.execute_region` (with tensor result). I think these bufferization pattern must be updated because they create new `to_tensor` ops.

**Comment by matthias-springer** [comment]
> Can you add extra unit tests for `scf.for` (with tensor iter_arg that has an encoding), `scf.if` and `scf.execute_region` (with tensor result). I think these bufferization pattern must be updated because they create new `to_tensor` ops.

Ideally, to make this PR not even larger, if an update to these patterns is indeed needed, you could update those patterns in a first, separate PR. I think not much has to change, you just have to take the tensor types from the input IR instead of inferring them in the op builder.

File: mlir/test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir (hunk 4 above)
Comment (christopherbate) [reply mlir/test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir:81]: Since having a `copy` generates a `memref.copy`, this meaning is to copy data between memory spaces, however `bufferization.alloc_tensor` only prints one type for both `copy` operand and the result. So we will need to update ASM format for `bufferization.alloc_tensor` as well. That will remove the need for the `tensor.cast`.

**Comment by aartbik** [comment]
lgtm for all sparse test signature changes

File: mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td (hunk 1 above)
Comment (christopherbate) [reply mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td]: Fixed

File: mlir/lib/Dialect/Bufferization/Transforms/Bufferize.cpp (hunk 2 above)
Comment (christopherbate) [reply mlir/lib/Dialect/Bufferization/Transforms/Bufferize.cpp]: I removed this. At the end of this materialization function there is ` llvm_unreachable("only tensor/memref input types supported");`, but if the input is `UnrankedMemRefType`, then it will still hit the `unreachable`.  Is that by design?

File: mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td (hunk 5 above)
Comment (christopherbate) [reply mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td:393]: done

File: mlir/include/mlir/Dialect/Bufferization/IR/Bufferization.h [hunk 6]
```diff
@@ -12,10 +12,16 @@
...
 #include "mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
 #include "mlir/Interfaces/CopyOpInterface.h"
 #include "mlir/Interfaces/DestinationStyleOpInterface.h"
 #include "mlir/Interfaces/InferTypeOpInterface.h"
 #include "mlir/Interfaces/SubsetOpInterface.h"
+#include "llvm/Support/Debug.h"
+
+namespace mlir::bufferization::detail {
+bool tensorTypesMatchUpToEncoding(Type lhs, Type rhs);
```
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/Bufferization/IR/Bufferization.h:23]: This is not used.

File: mlir/lib/Dialect/Bufferization/Transforms/Bufferize.cpp [hunk 7]
```diff
@@ -222,6 +222,13 @@ struct OneShotBufferizePass
             [](TensorType t) -> std::optional<Attribute> {
           return std::nullopt;
         };
+      } else if (useEncodingForMemorySpace) {
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/Bufferize.cpp:225]: Can you emit an error/assertion if both `mustInferMemorySpace` and `useEncodingForMemorySpace` are set?

**Review by matthias-springer — APPROVED** [review]

**Comment by srcarroll** [comment]
is this PR dead? wondering why it's been sitting here approved for months

**Comment by christopherbate** [comment]
It's not dead; I have just been pressed for time. Let me rebase and address Matthias' last comments this week.

**Comment by CoTinker** [comment]
Hi, this PR is useful for me, do you have any time to rebase it. Thanks.
